R/build-search-docs.R

```diff
@@ -65,3 +68,185 @@ url_node <- function(url) {
     paste0("<url><loc>", url, "</loc></url>")
   )
 }
+
+#' Build search index
+#'
+#' Build a JSON file encompassing all HTML pages, for use by the search script.
+#'
+#' @section YAML config:
+#' You can exclude some paths from the search index.
+#' Below we exclude the changelog from the search index.
+#'
+#' ```yaml
+#' search:
+#'   exclude: ['news/index.html']
+#' ```
+#' @section Local testing:
+#' Search won't work if you simply use pkgdown preview of the static files.
+#' You can use `servr::httw("docs")` instead.
+#'
+#'
+#' @inheritParams build_articles
+#' @export
+#'
+build_search <- function(pkg = ".",
+                         override = list()) {
+  pkg <- section_init(pkg, depth = 1L, override = override)
+  rule("Building search index")
+
+  paths <- fs::path_rel(
+    fs::dir_ls(pkg$dst_path, glob = "*.html", recurse = TRUE),
+    pkg$dst_path
+  )
+  paths <- paths[!paths %in% c("404.html", "articles/index.html", "reference/index.html")]
+
+  # user-defined exclusions
+  paths <- paths[!paths %in% pkg$meta$search$exclude]
+
+  index <- lapply(paths, file_search_index, pkg = pkg)
+  index <- unlist(index, recursive = FALSE)
+  index <- purrr::compact(index)
+  jsonlite::write_json(
+    index,
+    file.path(pkg$dst_path, "search.json"),
+    auto_unbox = TRUE
+  )
+}
+
+file_search_index <- function(path, pkg) {
+  html <- xml2::read_html(file.path(pkg$dst_path, path))
+  # Get page title
+  title <- xml2::xml_find_first(html, ".//meta[@property='og:title']") %>%
+    xml2::xml_attr("content")
+
+  # Get contents minus logo
+  node <- xml2::xml_find_all(html, ".//div[contains(@class, 'contents')]")
+  xml2::xml_remove(xml2::xml_find_first(node, ".//img[contains(@class, 'pkg-logo')]"))
+  sections <- xml2::xml_find_all(node, ".//*[contains(@class, 'section')]")
+
+  # Function for extracting all headers
+  get_h <- function(level, section) {
+    parents <- xml2::xml_parents(section)
+    if (length(parents) == 0) {
+      return("")
+    }
+    parents <- parents[!is.na(xml2::xml_attr(parents, "class"))]
+    h_section <- parents[grepl(paste0("section level", level), xml2::xml_attr(parents, "class"))]
+    h <- xml2::xml_contents(h_section)[is_heading(xml2::xml_contents(h_section))]
+    heading_text <- sub("^\\n", "", xml_text1(h))
+    if (grepl("pkg-version", xml2::xml_attr(h, "class"))) {
+      sprintf("<b>%s</b>", heading_text)
```

Comment:
hacky way to bolden the version number. :grimacing:

Reply:
See slack comment — I meant we should bolden the `<h1>` instead of the final term in the breadcrumbs.